File: kvstore/eclipse_java/kvstore/src/orbisoftware/msgblit/kvstore/unittest/UnitTest.java
package orbisoftware.msgblit.kvstore.unittest;

import orbisoftware.msgblit.kvstore.KVStore;
import orbisoftware.msgblit.kvstore.KVStore.DistType;

public class UnitTest {

	public static void main(String args[]) {

		KVStore database = new KVStore("", 25);

		// *********** Start of the value declarations ***********

		byte byteTest1 = 127;
		byte byteTest2 = 103;
		byte byteTest3 = -128;

		short shortTest1 = 32767;
		short shortTest2 = 12345;
		short shortTest3 = -32768;

		int intTest1 = Integer.MAX_VALUE;
		int intTest2 = 3029482;
		int intTest3 = Integer.MIN_VALUE;

		long longTest1 = Long.MAX_VALUE;
		long longTest2 = 1112222333;
		long longTest3 = Long.MIN_VALUE;

		float floatTest1 = Float.MAX_VALUE;
		float floatTest2 = 541.12312f;
		float floatTest3 = Float.MIN_VALUE;

		double doubleTest1 = Double.MAX_VALUE;
		double doubleTest2 = 619887.789;
		double doubleTest3 = Double.MIN_VALUE;

		boolean boolTest1 = false;
		boolean boolTest2 = true;

		String stringTest1 = "this is a test from java";
		String stringTest2 = "";

		byte bytesTest1[] = new byte[] { 5, 6, 7 };
		byte bytesTest2[] = new byte[] {};

		// *********** Start of the array declarations ***********

		byte byteArrayTest1[] = new byte[] { -49, -47, -45, -43, -41 };
		byte byteArrayTest2[] = new byte[] {};

		short shortArrayTest1[] = new short[] { 32, 34, 36, 38, 40 };
		short shortArrayTest2[] = new short[] {};

		int intArrayTest1[] = new int[] { Integer.MAX_VALUE, Integer.MIN_VALUE,
				90, 120, 150, 180, 210, 240, 270, 300 };
		int intArrayTest2[] = new int[] {};

		long longArrayTest1[] = new long[] { -1000, -2000, Long.MIN_VALUE,
				Long.MAX_VALUE, 10000, 5000, 1000 };
		long longArrayTest2[] = new long[] {};

		float floatArrayTest1[] = new float[] { Float.MIN_VALUE, -291.9391f,
				49912.291f, Float.MAX_VALUE };
		float floatArrayTest2[] = new float[] {};

		double doubleArrayTest1[] = new double[] { Double.MAX_VALUE,
				Double.MIN_VALUE, 0.0, Math.PI, Math.E };
		double doubleArrayTest2[] = new double[] {};

		boolean boolArrayTest1[] = new boolean[] { false, false, false, true,
				true, true };
		boolean boolArrayTest2[] = new boolean[] {};

		database.startReq();

		database.setInt8Value("pig", byteTest2, DistType.SHARED);
		database.setInt16Value("cat", shortTest2, DistType.SHARED);
		database.setInt32Value("horse", intTest2, DistType.SHARED);
		database.setInt64Value("tiger", longTest2, DistType.SHARED);
		database.setFloatValue("dog", floatTest2, DistType.SHARED);
		database.setDoubleValue("monkey", doubleTest2, DistType.SHARED);
		database.setBoolValue("bird", boolTest2, DistType.SHARED);
		database.setStringValue("fox", stringTest1, DistType.SHARED);
		database.setByteBufferValue("chicken", bytesTest1, DistType.SHARED);
		database.setInt8Array("frog", byteArrayTest2, DistType.SHARED);
		database.setInt16Array("fish", shortArrayTest1, DistType.SHARED);
		database.setInt32Array("turtle", intArrayTest1, DistType.SHARED);
		database.setInt64Array("leopard", longArrayTest1, DistType.SHARED);
		database.setFloatArray("lamb", floatArrayTest1, DistType.SHARED);
		database.setDoubleArray("mule", doubleArrayTest1, DistType.SHARED);
		database.setBoolArray("penguin", boolArrayTest1, DistType.SHARED);

		while (true) {

			System.out.println("");

			// *********** Start of the value tests ***********

			byte byteRead;
			byteRead = database.getInt8Value("pig");
			System.out.println(byteRead);

			short shortRead;
			shortRead = database.getInt16Value("cat");
			System.out.println(shortRead);

			int intRead;
			intRead = database.getInt32Value("horse");
			System.out.println(intRead);

			long longRead;
			longRead = database.getInt64Value("tiger");
			System.out.println(longRead);

			float floatRead;
			floatRead = database.getFloatValue("dog");
			System.out.println(floatRead);

			double doubleRead;
			doubleRead = database.getDoubleValue("monkey");
			System.out.println(doubleRead);

			boolean boolRead;
			boolRead = database.getBoolValue("bird");
			System.out.println(boolRead);

			String stringRead;
			stringRead = database.getStringValue("fox");
			System.out.println(stringRead + stringRead);

			byte bytesRead[];
			bytesRead = database.getByteBufferValue("chicken");
			System.out.println("Byte Buffer Test - Read " + bytesRead.length
					+ " bytes");
			System.out.print("[");
			for (int i = 0; i < bytesRead.length; i++)
				System.out.print(bytesRead[i] + " ");
			System.out.println("]");

			// *********** Start of the array tests ***********

			byte byteArrayRead[];
			byteArrayRead = database.getInt8Array("frog");
			if (byteArrayRead != null) {
				System.out.println("Byte Array Test - Read "
						+ byteArrayRead.length + " elements");
				System.out.print("[");
				for (int i = 0; i < byteArrayRead.length; i++)
					System.out.print(byteArrayRead[i] + " ");
				System.out.println("]");
			}

			short shortArrayRead[];
			shortArrayRead = database.getInt16Array("fish");
			if (shortArrayRead != null) {
				System.out.println("Short Array Test - Read "
						+ shortArrayRead.length + " elements");
				System.out.print("[");
				for (int i = 0; i < shortArrayRead.length; i++)
					System.out.print(shortArrayRead[i] + " ");
				System.out.println("]");
			}

			int intArrayRead[];
			intArrayRead = database.getInt32Array("turtle");
			if (intArrayRead != null) {
				System.out.println("Int Array Test - Read "
						+ intArrayRead.length + " elements");
				System.out.print("[");
				for (int i = 0; i < intArrayRead.length; i++)
					System.out.print(intArrayRead[i] + " ");
				System.out.println("]");
			}

			long longArrayRead[];
			longArrayRead = database.getInt64Array("leopard");
			if (longArrayRead != null) {
				System.out.println("Long Array Test - Read "
						+ longArrayRead.length + " elements");
				System.out.print("[");
				for (int i = 0; i < longArrayRead.length; i++)
					System.out.print(longArrayRead[i] + " ");
				System.out.println("]");
			}

			float floatArrayRead[];
			floatArrayRead = database.getFloatArray("lamb");
			if (floatArrayRead != null) {
				System.out.println("Float Array Test - Read "
						+ floatArrayRead.length + " elements");
				System.out.print("[");
				for (int i = 0; i < floatArrayRead.length; i++)
					System.out.print(floatArrayRead[i] + " ");
				System.out.println("]");
			}

			double doubleArrayRead[];
			doubleArrayRead = database.getDoubleArray("mule");
			if (doubleArrayRead != null) {
				System.out.println("Double Array Test - Read "
						+ doubleArrayRead.length + " elements");
				System.out.print("[");
				for (int i = 0; i < doubleArrayRead.length; i++)
					System.out.print(doubleArrayRead[i] + " ");
				System.out.println("]");
			}

			boolean boolArrayRead[];
			boolArrayRead = database.getBoolArray("penguin");
			if (boolArrayRead != null) {
				System.out.println("Bool Array Test - Read "
						+ boolArrayRead.length + " elements");
				System.out.print("[");
				for (int i = 0; i < boolArrayRead.length; i++)
					System.out.print(boolArrayRead[i] + " ");
				System.out.println("]");
			}

			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}
	}
}

File: kvstore/eclipse_cpp/kvstore/src/unittest/UnitTest.cpp
#include "KVStore.hpp"
#include <boost/thread.hpp>
#include <climits>
#include <stdint.h>
#include <inttypes.h>
#include <float.h>

static bool shutdown = false;

void sigint_handler(int sig) {

   shutdown = true;
}

int main(int argc, char *argv[]) {

   signal(SIGINT, sigint_handler);

   KVStore* database = new KVStore("", 15);

   // ***********  Start of the value declarations ***********

   int8_t byteTest1 = CHAR_MAX;
   int8_t byteTest2 = 105;
   int8_t byteTest3 = CHAR_MIN;

   int16_t shortTest1 = SHRT_MAX;
   int16_t shortTest2 = 24111;
   int16_t shortTest3 = SHRT_MIN;

   int32_t intTest1 = INT_MAX;
   int32_t intTest2 = 7543937;
   int32_t intTest3 = INT_MIN;

   int64_t longTest1 = LONG_LONG_MAX;
   int64_t longTest2 = 299490021;
   int64_t longTest3 = LONG_LONG_MIN;

   float floatTest1 = FLT_MAX;
   float floatTest2 = 76.234911;
   float floatTest3 = FLT_MIN;

   double doubleTest1 = DBL_MAX;
   double doubleTest2 = 973.23521;
   double doubleTest3 = DBL_MIN;

   bool boolTest1 = false;
   bool boolTest2 = true;

   std::string stringTest1 = "this is a test from c++";
   std::string stringTest2 = "";

   int8_t bytesTest1[] = { 7, 8, 9 };
   int8_t bytesTest2[] = { };

   // ***********  Start of the array declarations ***********

   int8_t byteArrayTest1[] = { 87, 89, 91, 93, 95, 97, 99, 101 };
   int8_t byteArrayTest2[] = { };

   int16_t shortArrayTest1[] = { 18, 14, 16, 18, 20, 22 };
   int16_t shortArrayTest2[] = { };

   int32_t intArrayTest1[] = { 500, 1000, 1500, 2000, INT_MIN, INT_MAX };
   int32_t intArrayTest2[] = { };

   int64_t longArrayTest1[] = { LONG_LONG_MAX, 4000, 6000, 8000, 10000,
         LONG_LONG_MIN };
   int64_t longArrayTest2[] = { };

   float floatArrayTest1[] =
         { FLT_MIN, -2312.343, 83411.1423, 9211.41, FLT_MAX };
   float floatArrayTest2[] = { };

   double doubleArrayTest1[] = { DBL_MIN, -7312.52, -3411.1423, 611.41,
         98234.22, DBL_MAX };
   double doubleArrayTest2[] = { };

   bool boolArrayTest1[] =
         { true, true, true, true, false, false, false, false };
   bool boolArrayTest2[] = { };

   database->startReq();

   database->setInt8Value("pig", byteTest2, KVStore::SHARED);
   database->setInt16Value("cat", shortTest2, KVStore::SHARED);
   database->setInt32Value("horse", intTest2, KVStore::SHARED);
   database->setInt64Value("tiger", longTest2, KVStore::SHARED);
   database->setFloatValue("dog", floatTest2, KVStore::SHARED);
   database->setDoubleValue("monkey", doubleTest2, KVStore::SHARED);
   database->setBoolValue("bird", boolTest2, KVStore::SHARED);
   database->setStringValue("fox", stringTest1, KVStore::SHARED);
   database->setByteBufferValue("chicken", &bytesTest1[0], 3, KVStore::SHARED);
   database->setInt8Array("frog", byteArrayTest1, 8, KVStore::SHARED);
   database->setInt16Array("fish", shortArrayTest1, 6, KVStore::SHARED);
   database->setInt32Array("turtle", intArrayTest1, 6, KVStore::SHARED);
   database->setInt64Array("leopard", longArrayTest1, 6, KVStore::SHARED);
   database->setFloatArray("lamb", floatArrayTest1, 5, KVStore::SHARED);
   database->setDoubleArray("mule", doubleArrayTest1, 6, KVStore::SHARED);
   database->setBoolArray("penguin", boolArrayTest1, 8, KVStore::SHARED);

   while (true) {

      int bytesRead, numberElements;
      std::cout << std::endl;

      // ***********  Start of the value tests ***********

      int8_t byteRead;
      byteRead = database->getInt8Value("pig");
      std::cout << (int) byteRead << std::endl;

      short shortRead;
      shortRead = database->getInt16Value("cat");
      std::cout << shortRead << std::endl;

      int intRead;
      intRead = database->getInt32Value("horse");
      std::cout << intRead << std::endl;

      int64_t longRead;
      longRead = database->getInt64Value("tiger");
      std::cout << longRead << std::endl;

      float floatRead;
      floatRead = database->getFloatValue("dog");
      std::cout << floatRead << std::endl;

      double doubleRead;
      doubleRead = database->getDoubleValue("monkey");
      std::cout << doubleRead << std::endl;

      bool boolRead;
      boolRead = database->getBoolValue("bird");
      std::cout << boolRead << std::endl;

      std::string stringRead;
      stringRead = database->getStringValue("fox");
      std::cout << stringRead << stringRead << std::endl;

      int8_t byteBuffer[10];
      bytesRead = database->getByteBufferValue("chicken", &byteBuffer[0]);
      if (bytesRead > 0) {
         std::cout << "Byte Buffer Test - Read " << bytesRead << " bytes"
               << std::endl;
         std::cout << "[";
         for (int i = 0; i < bytesRead; i++)
            std::cout << (int) byteBuffer[i] << " ";
         std::cout << "]" << std::endl;
      }

      // ***********  Start of the array tests ***********
      int8_t byteArray[15];
      numberElements = database->getInt8Array("frog", byteArray);
      if (numberElements > 0) {
         std::cout << "Byte Array Test - Read " << numberElements << " elements"
               << std::endl;
         std::cout << "[";
         for (int i = 0; i < numberElements; i++)
            std::cout << (int) byteArray[i] << " ";
         std::cout << "]" << std::endl;
      }

      int16_t shortArray[10];
      numberElements = database->getInt16Array("fish", shortArray);
      if (numberElements > 0) {
         std::cout << "Short Array Test - Read " << numberElements
               << " elements" << std::endl;
         std::cout << "[";
         for (int i = 0; i < numberElements; i++)
            std::cout << (short) shortArray[i] << " ";
         std::cout << "]" << std::endl;
      }

      int32_t intArray[15];
      numberElements = database->getInt32Array("turtle", intArray);
      if (numberElements > 0) {
         std::cout << "Int Array Test - Read " << numberElements << " elements"
               << std::endl;
         std::cout << "[";
         for (int i = 0; i < numberElements; i++)
            std::cout << (int) intArray[i] << " ";
         std::cout << "]" << std::endl;
      }

      int64_t longArray[15];
      numberElements = database->getInt64Array("leopard", longArray);
      if (numberElements > 0) {
         std::cout << "Long Array Test - Read " << numberElements << " elements"
               << std::endl;
         std::cout << "[";
         for (int i = 0; i < numberElements; i++)
            std::cout << (long long) longArray[i] << " ";
         std::cout << "]" << std::endl;
      }

      float floatArray[15];
      numberElements = database->getFloatArray("lamb", floatArray);
      if (numberElements > 0) {
         std::cout << "Float Array Test - Read " << numberElements
               << " elements" << std::endl;
         std::cout << "[";
         for (int i = 0; i < numberElements; i++)
            std::cout << (float) floatArray[i] << " ";
         std::cout << "]" << std::endl;
      }

      double doubleArray[15];
      numberElements = database->getDoubleArray("mule", doubleArray);
      if (numberElements > 0) {
         std::cout << "Double Array Test - Read " << numberElements
               << " elements" << std::endl;
         std::cout << "[";
         for (int i = 0; i < numberElements; i++)
            std::cout << (double) doubleArray[i] << " ";
         std::cout << "]" << std::endl;
      }

      bool boolArray[15];
      numberElements = database->getBoolArray("penguin", boolArray);
      if (numberElements > 0) {
         std::cout << "Bool Array Test - Read " << numberElements << " elements"
               << std::endl;
         std::cout << "[";
         for (int i = 0; i < numberElements; i++)
            std::cout << (bool) boolArray[i] << " ";
         std::cout << "]" << std::endl;
      }

      if (shutdown) {
         database->shutdownReq();

         while (!database->threadIsComplete())
            boost::this_thread::sleep(boost::posix_time::milliseconds(200));

         abort();
      }

      boost::this_thread::sleep(boost::posix_time::milliseconds(1000));
   }

   return 0;
}
